Nonholonomic Equations of Motion

moorepants opened the pull request.

moorepants commented:
@jtheinen and @Peter230655 this one is ready for review. I'm going to publish it by the end of the day Monday. Here is the animation:

https://user-images.githubusercontent.com/276007/168414525-a377c639-ce41-4fa6-91f3-2461f7abc662.mp4

Peter230655 commented:
1.
You write:


*We will address the nonholonomic equations of motion first, as they are
slightly easier to deal with.*
I am looking forward to your section on holonomic constraints.
I thought, they are MUCH harder as they are generally nonlinear.

2. Snakeboard Equations of Motion

Would it make sense to repeat the sketch of the snakeboard here?
Maybe it would help to more easily recall the definitions of the frames and
the generalized coordinates?

3.
A part of your code reads


O.set_vel(N, 0)Ao.vel(N)Bo.v2pt_theory(Ao, N, A)Co.v2pt_theory(Ao, N, A);


Is Ao.vel(N) needed here?
In the original code you used it to print the speed of Ao

4.
You write:

*In our case, the dependent generalized speeds are only a function of one
independent generalized speed, **𝑢3*u3
It took me a while to (hopefully) understand, why this must be so: A small
change in q4, q5, caused by u4, u5 intuitively does not affect u1, u2.
No idea, whether such reasoning is useful for students to do.

5. Warning

These chained replacements work from left to right, so to speak, which is
‚contrary‘ to normal Python rules - or am I getting something wrong?

6. Animation

It is absolutely *GREAT*!!
I could not find, where you set the colors of the lines which trace the
motion of the wheels. Is this set automatically, or did I miss it?

7. Calculating Dependent Speeds

ax.legend(['$u_1$', '$u_2$'])


Maybe you could put a semicolon ( ; ) after this line to avoid the printout
<matplotlib.legend…..> as you explained earlier?



On Sat 14. May 2022 at 13:52 Jason K. Moore ***@***.***>
wrote:

> @jtheinen <https://github.com/jtheinen> and @Peter230655
> <https://github.com/Peter230655> this one is ready for review. I'm going
> to publish it by the end of the day Monday. Here is the animation:
>
>
> https://user-images.githubusercontent.com/276007/168414525-a377c639-ce41-4fa6-91f3-2461f7abc662.mp4
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/moorepants/learn-multibody-dynamics/pull/75#issuecomment-1126657097>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AT5MQUQAWRFGGJFENLUPCTLVJ5ERTANCNFSM5VWBO5BA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
-- 
Best regards,

Peter Stahlecker

moorepants commented:
> 1. You write: *We will address the nonholonomic equations of motion first, as they are slightly easier to deal with.* I am looking forward to your section on holonomic constraints. I thought, they are MUCH harder as they are generally nonlinear.

"Slightly" could slightly be an understatement :). But maybe not too bad. It'll be in the next chapter.

> 2. Snakeboard Equations of Motion
> 
> Would it make sense to repeat the sketch of the snakeboard here?
> Maybe it would help to more easily recall the definitions of the frames and
> the generalized coordinates?

Yes, will do.

> 3.
> A part of your code reads
> O.set_vel(N, 0)Ao.vel(N)Bo.v2pt_theory(Ao, N, A)Co.v2pt_theory(Ao, N, A);
> Is Ao.vel(N) needed here?
> In the original code you used it to print the speed of Ao

Removed.

> 4.
> You write:
> 
> *In our case, the dependent generalized speeds are only a function of one
> independent generalized speed, **𝑢3*u3
> It took me a while to (hopefully) understand, why this must be so: A small
> change in q4, q5, caused by u4, u5 intuitively does not affect u1, u2.
> No idea, whether such reasoning is useful for students to do.

I added a little exercise question (but without an answer) to provoke the though experiment.

> 5. Warning
> 
> These chained replacements work from left to right, so to speak, which is
> ‚contrary‘ to normal Python rules - or am I getting something wrong?

```python
a = something
b = a.xreplace(x)
c = b.xreplace(y)
```
is the same as doing:

``python
a = something
c = a.xreplace(x).xreplace(y)
```

This chaining of commands works from left to right: x is substituted first then y.


> 6. Animation
> 
> It is absolutely *GREAT*!!
> I could not find, where you set the colors of the lines which trace the
> motion of the wheels. Is this set automatically, or did I miss it?

Glad you liked it. The colors happen automatically with matplolib. Every new line you add to a plot will cycle through matplotlib's default color list. So I got a rainbow effect for free.

> 7. Calculating Dependent Speeds
> 
> ax.legend(['$u_1$', '$u_2$'])

Fixed.

moorepants commented on nonholonomic-eom.rst:
```diff
@@ -0,0 +1,660 @@
+The snakeboard model, as described, has no generalized active forces because
+there are no contributing external forces acting on the system, so we only need
+to generate the generalized inertia forces :math:`\bar{F}_r^*`. We now then
+calculate the velocities we will need to form :math:`\bar{F}_r^*` and make sure
```
These should be the tilde versions to represent nonholonomic.

jtheinen reviewed (commented):
Good that you explained all the code thoroughly, all the python is clear now. 

During reading I sometimes find it hard to know at what step I am of the process. For you it is obvious what steps you need to take to reach the goal and in which order. I think it would be very useful to have a summarized/general 'step plan' for the whole process from problem to answer. such as:
1. Generalized/independent speeds
2. Frames
3. declare positions/speeds of important points in terms of gen. coord/speeds
4. ...
5. numerical integration
6. simulation

Maybe this could be a separate page on the website. "practical guide"

moorepants commented on nonholonomic-eom.rst:
```diff
@@ -0,0 +1,661 @@
+   x = sm.Matrix([q1, q2, q3, q4, q5, u3, u4, u5])
+   eval_ur = sm.lambdify((x, p), ur_sol)
+
+   ur_vals = eval_ur(np.transpose(xs), np.transpose(np.tile(p_vals, (600, 1)))).squeeze()
```
Yeah, doing this probably is not good for this text. It's using NumPy's broadcasting features to avoid having to write a loop.

https://numpy.org/doc/stable/user/basics.broadcasting.html

I'll change it to a loop.

moorepants replied:
Changed.

moorepants commented on nonholonomic-eom.rst:
```diff
@@ -0,0 +1,661 @@
+       cl_path.set_data(coords[:i, 0, 0], coords[:i, 1, 0])
+       cr_path.set_data(coords[:i, 0, 2], coords[:i, 1, 2])
+       bl_path.set_data(coords[:i, 0, 6], coords[:i, 1, 6])
+       br_path.set_data(coords[:i, 0, 7], coords[:i, 1, 7])
```
I could ask them to do that for the homework!

moorepants commented on nonholonomic-eom.rst:
```diff
@@ -0,0 +1,661 @@
+
+.. jupyter-execute::
+
+   I_A_Ao = I*me.outer(A.z, A.z)
```
No, it isn't needed in the way I've set things up. I only need to give the central moment of inertia of each body an the inertia torques are formed separately. In some sense, the parallel axis theorem will implicitly be applied.

moorepants commented on nonholonomic-eom.rst:
```diff
@@ -0,0 +1,661 @@
+
+In chapters, :ref:`Holonomic Constraints` and :ref:`Nonholonomic Constraints`,
+I introduced two types of constraints: holonomic (configuration) constraints
+and nonholonomic (motion) constraints, respectively. Holonomic constraints are
```
Removed "respectively".

moorepants commented:
> I think it would be very useful to have a summarized/general 'step plan' for the whole process from problem to answer.

I added headings for this in this chapter that may help some. But I like the idea of having a step by step guide as a separate document.

moorepants commented:
This one is live! https://moorepants.github.io/learn-multibody-dynamics/nonholonomic-eom.html
